pkg/algorithm/hived_algorithm.go

```diff
@@ -149,49 +152,92 @@ func (h *HivedAlgorithm) Schedule(
 	pod *core.Pod,
 	suggestedNodes []string,
 	phase internal.SchedulingPhase) internal.PodScheduleResult {
+
 	h.algorithmLock.Lock()
 	defer h.algorithmLock.Unlock()
 
-	klog.Infof("[%v]: Scheduling pod...", internal.Key(pod))
+	klog.Infof("[%v]: Scheduling pod in %v phase...", internal.Key(pod), phase)
 	s := internal.ExtractPodSchedulingSpec(pod)
-	// gpu number -> a set of pods -> a set of GPUs of each pod
-	groupPhysicalPlacement := groupPhysicalPlacement{}
-	groupVirtualPlacement := groupVirtualPlacement{}
-	podIndex := int32(0)
 	suggestedNodeSet := common.NewSet()
 	for _, n := range suggestedNodes {
 		suggestedNodeSet.Add(n)
 	}
-
-	group := h.allocatedAffinityGroups[s.AffinityGroup.Name]
-	if group == nil {
+	var (
+		groupPhysicalPlacement groupPhysicalPlacement // GPU number -> a set of pods -> a set of GPUs of each pod
+		groupVirtualPlacement  groupVirtualPlacement  // GPU number -> a set of pods -> a set of GPUs of each pod
+		preemptionVictims      map[string]common.Set  // node -> pods
+		nodesNotInSuggested    common.Set             // nodes in physical placement that are not within suggested nodes
+		podIndex               int32                  // index of current pod among those of the same GPU number in the group, 0 by default
+	)
+
+	if g := h.affinityGroups[s.AffinityGroup.Name]; g != nil {
+		// state of g can be either Allocated or Preempting
+		if g.state == groupAllocated {
+			klog.Infof("[%v]: Pod affinity group is already allocated: %v", internal.Key(pod), s.AffinityGroup.Name)
+			groupPhysicalPlacement = g.physicalGpuPlacement
+			groupVirtualPlacement = g.virtualGpuPlacement
+			if podIndex = getNewPodIndex(g.allocatedPods[s.GpuNumber]); podIndex == -1 {
+				panic(internal.NewBadRequestError(fmt.Sprintf(
+					"Requesting more pods than the configured number for %v GPUs (%v pods) in affinity group %v",
+					s.GpuNumber, g.totalPodNums[s.GpuNumber], s.AffinityGroup.Name)))
+			}
+		} else { // groupPreempting
+			klog.Infof("[%v]: Pod affinity group is preempting others: %v", internal.Key(pod), s.AffinityGroup.Name)
+			nodesNotInSuggested = collectNodesNotSuggested(g.physicalGpuPlacement, suggestedNodeSet)
+			if phase == internal.PreemptingPhase && !nodesNotInSuggested.IsEmpty() {
+				// If we find a preempting group's placement is not fully within suggested nodes, we should cancel
+				// the preemption so as to reschedule it to other places. We should do this only in Preempting phase
+				// because the suggested nodes in Filtering phase does not consider preemption.
+				klog.Infof("[%v]: Placement not fully within Preempting-phase suggested nodes, "+
+					"deleting preempting affinity group %v", internal.Key(pod), g.name)
```

Comment:
>, internal.Key(pod), g.name) [](start = 44, length = 28)

Also print these nodesNotInSuggested?

Reply:
Align the cancel log format, such as:
"Affinity group %v preemption canceled due to its placement is no longer fully within suggested nodes"

"Affinity group %v preemption canceled due to it is further preempted by another higher priority affinity group"

"Affinity group %v preemption canceled due to it is deleted"

etc



---
In reply to: [399953733](https://github.com/microsoft/hivedscheduler/pull/8#discussion_r399953733) [](ancestors = 399953733)